File: src/learning_empty_world.cpp
#include "turtlebot3_rl/Toolbox.h"
#include "turtlebot3_rl/ReinforcementLearning.h"
#include "turtlebot3_rl/TurtleMove.h"
#include <iostream>
#include <fstream>
#include <iomanip>

int main(int argc, char **argv) {

    int   n_actions = 3;
    float epsilon   = 0.9;

    // rl object :
    QLearning rl(n_actions, epsilon);

    // create actions :
    int *actions = rl.create_actions(n_actions);

    // state space info :
    struct stateInfo x;
    x.min = -5.0;
    x.max = +5.0;
    x.n   =  11;
    
    struct stateInfo y;
    y.min = -5.0;
    y.max = +5.0;
    y.n   =  11;

    struct stateInfo theta;
    theta.min = - _PI_NUMBER_;
    theta.max =   _PI_NUMBER_;
    theta.n   =  9;

    // create states space :
    float **all_states = rl.create_states(x, y, theta);
    std::size_t n_all_states = x.n * y.n * theta.n;
    std::size_t n_columns = 3;

    // pairing states and actions :
    float **state_action_pairs = rl.create_state_action_pairs(all_states, actions);
    std::size_t n_state_action_pairs   = n_all_states * n_actions;
    std::size_t n_columns_action_pairs = n_columns + 1;

    // learning params :
    int   n_episodes = 50000;
    float gamma      = 0.990;
    float alpha      = 0.100;

    // goal selection :
    float  goal_state[] = {4.0, -4.0};
    float *goal = goal_state;

    // initialize polices :
    int *policy = rl.get_random_policy(n_all_states);

    // initialize q table :
    float *Qtable = rl.zeros_init_Qtable(n_state_action_pairs);

    // initialize q table window for calculate q table in state s for all actions :
    float *Qtable_window = new float[n_actions];

    // init current and next state pointers :
    float *state = new float[n_columns];
    float *state_next = new float[n_columns];

    // init state index with None :
    int state_idx = -1; 

    // init action with None :
    int action = -1;

    // init a rand var with None :
    float rand = -1.0;

    // init reward per episode (befor starts learning, reward is zero):
    float reward = 0.0;

    // init paired state and action in each episode :
    float *state_action_paired = new float[n_columns_action_pairs];
    
    // init state_action_paired_idx for calculat the best idx that maximized q table :
    int  state_action_paired_idx;

    // init state_action_paired_next_idx for calculat the best idx of next in q table :
    int *state_action_paired_next_idx = new int[n_actions];

    // max finder object :
    maxFind max_of_Qtable;

    // reward per episode :
    float rpe = 0.0;

    int i = 0;
    // learning loop :
    for(int e=0; e<n_episodes; e++) {
        
        // select a random state :
        int random_row = rl.randGenerator.randInteger(0, n_all_states - 1);
        for(i=0; i<n_columns; i++) {
            *(state + i) = *(*(all_states + random_row) + i);
        }
        
        // reset rpe :
        rpe = 0.0;
        while(1) {
            // find random selected state in all_states :
            state_idx = ismember<float>(state, all_states, n_all_states, n_columns);
            // std::cout << state_idx << std::endl;
            
            // select an action :
            rand = rl.randGenerator.randUniform(0.0, 1.0);
            if(rand > epsilon) {
                // select an action from optimal policy :
                action = *(policy + state_idx);
            }
            else {
                // select a random action :
                action = rl.get_random_action(rl.actions);
            }

            // doing selected action and get reward and new state :
            rl.do_action(state, state_next, action);

            // calculate reward :
            reward = rl.get_reward(all_states, state_next, state, goal, action);

            // create state_action_paired at this state for all actions:
            for(i=0; i<n_columns; i++) {
                *(state_action_paired + i) = *(state + i);
            }
            *(state_action_paired + n_columns) = action;

            // find state_action_paired_idx that maximized q table :
            state_action_paired_idx = ismember<float>(state_action_paired, state_action_pairs, n_state_action_pairs, n_columns_action_pairs);
            *state_action_paired_next_idx = ismember<float>(state_next, state_action_pairs, n_state_action_pairs, n_columns);
            // std::cout << *state_action_paired_next_idx << std::endl;
            
            for(i=0; i<n_actions; i++) {
                *(Qtable_window + i) = Qtable[*(state_action_paired_next_idx) + i];
            }

            max_of_Qtable.maxFunc(Qtable_window, n_actions);
            Qtable[state_action_paired_idx] = Qtable[state_action_paired_idx] + alpha * (reward + gamma * max_of_Qtable.maxVal - Qtable[state_action_paired_idx]);

            // policy improvement :
            rl.improve_policy(Qtable, n_state_action_pairs, policy);
        
            // update states :
            *(state + 0) = *(state_next + 0);
            *(state + 1) = *(state_next + 1);
            *(state + 2) = *(state_next + 2);

            // some of rewards :
            rpe += reward;

            // check reach goal state :
            if(*(state + 0) == *(goal + 0) && *(state + 1) == *(goal + 1)) {
              break;
            }
        }
        std::cout << "Episode : " << e << "\t" << "Reward per Episode : " << rpe << std::endl;
        epsilon = epsilon * 0.97;
    }

    // create a text file for log learning process :
    std::ofstream learning_data;

    // file path :
    learning_data.open("/home/ali/catkin_ws/src/turtlebot3_rl/LogData/data_empty_world.txt");
    
    // log algorithm params :
    learning_data << "Gamma     = " << gamma << " , " << "alpha    = " << alpha << " , " << "episodes            = " << n_episodes << std::endl;
    learning_data << "n actions = " << n_actions << "    , " << "n states = " << n_all_states << " , " << "n state action pair = " << n_state_action_pairs << std::endl;
    learning_data << std::endl;
    learning_data << "row" << "\t \t " << "x" << "\t \t"  << " y" << "\t \t"  << "theta" << "\t \t"  << "actions" << "\t \t"  << "best action" << "\t \t" << "Q Table" << std::endl;
    
    // log q table and optimal policy :
    int index=0; // for log optimal policy
    for(int i=0; i<n_state_action_pairs; i++) {
        learning_data << i << "\t \t" << *(*(state_action_pairs + i) + 0) << "\t \t" << *(*(state_action_pairs + i) + 1) << "\t \t" << std::setprecision(4) << *(*(state_action_pairs + i) + 2) << "\t \t" << *(*(state_action_pairs + i) + 3) << "\t \t" << *(policy + index) << "\t \t \t" << std::setprecision(4) << *(Qtable + i) << std::endl;
        if(((i + 1) % n_actions) == 0) {index++;}
    }

    // save and close file:
    learning_data.close();

    return 0;
}

File: MATLAB Scripts/learning_empty_world.m
clc
clear
close all

%% Create Spaces and Pairs

% x space :
x.min = -2.5;
x.max = +2.5;
x.n   = 11;
x.space_precise = (x.max - x.min) / (x.n - 1);
x.vec = linspace(x.min, x.max, x.n);

% y space :
y.min = -2.5;
y.max = +2.5;
y.n   = 11;
y.space_precise = (y.max - y.min) / (y.n - 1);
y.vec = linspace(y.min, y.max, y.n);

% theta space :
theta.min = -pi;
theta.max = +pi;
theta.n   = 9;
theta.space_precise = (theta.max - theta.min) / (theta.n - 1);
theta.vec = linspace(theta.min, theta.max, theta.n);

% create state space :
[first_col, second_col, third_col] = ndgrid(x.vec, y.vec, theta.vec);
all_states = [first_col(:), second_col(:), third_col(:)];
n_all_states = size(all_states, 1);
n_columns = size(all_states, 2);


%% Declare Actions and State-Action Pairs

% possible actions :
n_actions = 3;
actions = [0; 1; 2];

% create state action pairs :
[first_col, second_col, third_col, fourth_col] = ndgrid(x.vec, y.vec, theta.vec, actions);
state_action_pairs = [first_col(:), second_col(:), third_col(:), fourth_col(:)];
n_state_action_pairs = size(state_action_pairs, 1);
n_columns_action_pairs = size(state_action_pairs, 2);

%% Initialization

% goal point :
goal = [-0.5, 2.5];

% for store optimal policy :
policy = zeros(n_all_states, 1);

% for store Q Table :
QTable = zeros(n_state_action_pairs, 1);

% Q learning algorithm params :
n_episodes = 30000;
epsilon    = 0.9;
gamma      = 0.99;
alpha      = 0.1;

%% Q learning Algorithm

% start time :
tic;

% learning loop :
for e=1:n_episodes
    % select a random row :
    random_row = randi(n_all_states - 1);

    % state in episode :
    state = all_states(random_row, :);

    % reset rpe :
    rpe = 0;

    % main loop :
    while true
        state_idx = find(ismember(all_states, state, 'rows'));

        % epsilonilon greedy :
        if rand > epsilon
            action = policy(state_idx);
        else
            action = actions(randi(n_actions));
        end

        % do selected action :
        state_next = do_action(state, action, x, y, theta);

        % get reward and check new state status :
        [reward, state_next] = get_reward(all_states, state_next, state, goal);

        % find state_action_paired_idx that maximized q table :
        state_action_paired_idx = find(ismember(state_action_pairs, [state, action], 'rows'));
        state_action_paired_next_idx = find(ismember(state_action_pairs(:, 1:3), state_next, 'rows'));
        
        % update QTable :
        max_QTable = max(QTable(state_action_paired_next_idx)) - QTable(state_action_paired_idx);
        QTable(state_action_paired_idx) = QTable(state_action_paired_idx) + alpha * (reward + gamma * max_QTable);
        
        % improve policy :
        [~, argmax] = max(QTable(state_action_paired_next_idx));
        policy(state_action_paired_idx) = argmax - 1;
        % for i = 1:n_all_states
        %     [~, argmax] = max(QTable(1+n_actions*(i-1):n_actions*i));
        %     policy(i) = argmax;
        % end

        % update state :
        state = state_next;

        % reward per episode :
        rpe = rpe + reward;

        % check reaching condition :
        if all(state(1:2) == goal)
            break;
        end
    end
    disp(['Episode : ', num2str(e), ' Reward Per Episode : ', num2str(rpe)]);
    epsilon = epsilon * 0.97;
end

% time duration :
toc;

%% Local Functions

function state_next = do_action(state, action, x_space, y_space, theta_space)

    % exract states :
    x     = state(1);
    y     = state(2);
    theta = state(3);

    % space precises :
    x_space_precise     = x_space.space_precise;
    y_space_precise     = y_space.space_precise;
    theta_space_precise = theta_space.space_precise;

    epsilon = 0.01;
    if abs(theta + pi) < epsilon
        r1 = - 1.0;
        r2 =   0.0;
    elseif abs(theta + 3 * pi / 4.0) < epsilon
        r1 = - 1.0;
        r2 = - 1.0;
    elseif abs(theta + pi / 2.0) < epsilon
        r1 =   0.0;
        r2 = - 1.0;
    elseif abs(theta + pi / 4.0) < epsilon
        r1 =   1.0;
        r2 = - 1.0;
    elseif abs(theta) < epsilon
        r1 =   1.0;
        r2 =   0.0;
    elseif abs(theta - pi / 4.0) < epsilon
        r1 =   1.0;
        r2 =   1.0;
    elseif abs(theta - pi / 2.0) < epsilon
        r1 =   0.0;
        r2 =   1.0;
    elseif abs(theta - 3 * pi / 4.0) < epsilon
        r1 = - 1.0;
        r2 =   1.0;
    elseif abs(theta - pi) < epsilon
        r1 = - 1.0;
        r2 =   0.0;
    else
        disp("Are You Joking?!");
    end

    switch (action)
        case 0
            x_new = x + x_space_precise * r1;
            y_new = y + y_space_precise * r2;
            theta_new = theta;
    
        case 1
            x_new = x;
            y_new = y;
            theta_new = theta - theta_space_precise;
    
        case 2
            x_new = x;
            y_new = y;
            theta_new = theta + theta_space_precise;

        otherwise
            x_new = x;
            y_new = y;
            theta_new = theta;
    end

    % update state :
    state_next = zeros(1, 3);
    state_next(1) = x_new;
    state_next(2) = y_new;
    state_next(3) = theta_new;
end

function [reward, new_state] = get_reward(all_states, state_next, state, goal)
    state_status = find(ismember(all_states, state_next, 'rows')); %#ok
    new_state = state_next;
    if ~isempty(state_status)
        if all(state_next(1:2) == goal)
          reward = +10.0;
        else
          reward = -1.0;
        end
    else
        new_state = state;
        reward = -10.0;
    end
end